### PR TITLE
Rand function for complex dtype

### PR DESCRIPTION
**Pull request by kostekIV**
Address #34380

### CODE REVIEW

**Comment by kostekIV** [comment]
Hi @anjali411 can you take look at this? I can't edit reviewers, also not sure why ci stopped at only 6 checks. Side note didn't run into issue with `view_complex_as_float` in header file so I didn't make it inline.

**Comment by anjali411** [comment]
Hi @kostekIV thanks for working on this! :D I wanted to have randn work for complex tensors on cuda so I added a header here https://github.com/pytorch/pytorch/pull/35056. Once this PR is landed, do you mind updating your PR and also adding rand support for complex on cuda?

**Comment by kostekIV** [comment]
Sure @anjali411 I will do that. Correct me if I am wrong but doesn't `rand` function in `aten/src/ATen/native/TensorFactories.cpp` already is for both cpu and cuda, so the change in this pr (after syncing it with #35056)  is sufficient for it to work on both cpu and cuda?

**Comment by anjali411** [comment]
> Sure @anjali411 I will do that. Correct me if I am wrong but doesn't `rand` function in `aten/src/ATen/native/TensorFactories.cpp` already is for both cpu and cuda, so the change in this pr (after syncing it with #35056) is sufficient for it to work on both cpu and cuda?

oh yeah you are right! randn calls into uniform_ which has separate cpu and cuda calls so we don't need to worry about it. This PR looks good overall. I'll let you know once I land https://github.com/pytorch/pytorch/pull/35056 and then we can land this PR after removing the view_complex_as_float header specific changes. thanks for working on this!! :D

**Comment by kostekIV** [comment]
@anjali411 updated

**Review by anjali411 — APPROVED** [review]
LGTM :) thanks for the awesome work!

File: aten/src/ATen/native/TensorFactories.cpp [hunk 1]
```diff
@@ -475,6 +476,12 @@ Tensor rand(IntArrayRef size, const TensorOptions& options) {
 
 Tensor rand(IntArrayRef size, Generator* generator, const TensorOptions& options) {
   auto result = at::empty(size, options);
+  if (result.is_complex()) {
+    auto float_tensor = at::native::view_complex_as_float(result);
+    float_tensor.uniform_(0, 1, generator);
+    // we want to return complex tensor not underlaying float tensor.
```
Comment (anjali411) [inline aten/src/ATen/native/TensorFactories.cpp]: nit: typo- not underlying

Comment (kostekIV) [reply aten/src/ATen/native/TensorFactories.cpp]: oops, fixed

**Comment by anjali411** [comment]
@kostekIV two tests seem to be failing due to this PR. can you look into it?

**Comment by kostekIV** [comment]
@anjali411 I am honestly at lost here, I was looking for an error and couldn't find what is causing those failures here. I looked into other prs and few of the already merged ones had this 2 failures, for example [this](https://github.com/pytorch/pytorch/pull/35242) had `torch_xla/csrc/aten_xla_type_default.cpp:11460:8: error: no matching member function for call to 'impl_unboxedOnlyKernel'` and [this](https://github.com/pytorch/pytorch/pull/35282) had `AssertionError: 11 not less than or equal to 1e-05`. Not sure how to approach this. Thank you for your help :)

**Pull request closed without merging**